sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/SearchClientBuilder.java

```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Creates a {@link SearchClient} based on options set in the Builder. Every time {@code buildClient()} is
+     * called a new instance of {@link SearchClient} is created.
+     * <p>
+     * If {@link #pipeline(HttpPipeline) pipeline} is set, then only the {@code pipeline}, {@link #endpoint(String)
+     * endpoint}, and {@link #indexName(String) indexName} are used to create the {@link SearchClient client}.
+     * All other builder settings are ignored.
+     *
+     * @return A SearchClient with the options set from the builder.
+     * @throws NullPointerException If {@code indexName} or {@code endpoint} are {@code null}.
+     */
+    public SearchClient buildClient() {
+        return new SearchClient(buildAsyncClient());
+    }
+
+    /**
+     * Creates a {@link SearchAsyncClient} based on options set in the Builder. Every time
+     * {@code buildAsyncClient()} is called a new instance of {@link SearchAsyncClient} is created.
+     * <p>
+     * If {@link #pipeline(HttpPipeline) pipeline} is set, then only the {@code pipeline}, {@link #endpoint(String)
+     * endpoint}, and {@link #indexName(String) indexName} are used to create the {@link SearchAsyncClient client}.
+     * All other builder settings are ignored.
+     *
+     * @return A SearchClient with the options set from the builder.
+     * @throws NullPointerException If {@code indexName} or {@code endpoint} are {@code null}.
+     */
+    public SearchAsyncClient buildAsyncClient() {
+        Objects.requireNonNull(indexName, "'indexName' cannot be null.");
+        Objects.requireNonNull(endpoint, "'endpoint' cannot be null.");
+
+        SearchServiceVersion buildVersion = (serviceVersion == null)
+            ? SearchServiceVersion.getLatest()
+            : serviceVersion;
+
+        if (httpPipeline != null) {
+            return new SearchAsyncClient(endpoint, indexName, buildVersion, httpPipeline);
+        }
+
+        Configuration buildConfiguration = (configuration == null)
+            ? Configuration.getGlobalConfiguration()
+            : configuration;
+
+        // Closest to API goes first, closest to wire goes last.
+        final List<HttpPipelinePolicy> httpPipelinePolicies = new ArrayList<>();
+        httpPipelinePolicies.add(new AddHeadersPolicy(headers));
+        httpPipelinePolicies.add(new RequestIdPolicy());
+
+        HttpPolicyProviders.addBeforeRetryPolicies(httpPipelinePolicies);
+        httpPipelinePolicies.add(retryPolicy == null ? new RetryPolicy() : retryPolicy);
+
+        httpPipelinePolicies.add(new AddDatePolicy());
+        if (keyCredential != null) {
+            this.policies.add(new AzureKeyCredentialPolicy(API_KEY, keyCredential));
+        }
+        httpPipelinePolicies.addAll(this.policies);
+
+        HttpPolicyProviders.addAfterRetryPolicies(httpPipelinePolicies);
+
+        httpPipelinePolicies.add(new UserAgentPolicy(httpLogOptions.getApplicationId(), clientName, clientVersion,
```

Comment:
Do we normally add the `User-Agent` at the end of the pipeline? I thought this normally was added when setting the request ID since it'll be static throughout the request/requests (if retried) being made.